src/Workspaces/Core/Portable/Workspace/WorkspaceEventMap.cs

```diff
@@ -0,0 +1,129 @@
+            // disposal would have called this method.
+            Debug.Assert(originalHandlers != null);
+
+            if (originalHandlers != null)
+            {
+                var newHandlers = originalHandlers.RemoveHandler(handlerAndOptions);
+                Debug.Assert(originalHandlers != newHandlers);
+
+                _eventTypeToHandlerSet[eventType] = newHandlers;
+            }
+        }
+    }
+
+    public EventHandlerSet GetEventHandlerSet(WorkspaceEventType eventType)
+    {
+        using (_guard.DisposableWait())
+        {
+            return _eventTypeToHandlerSet.TryGetValue(eventType, out var handlers)
+                ? handlers
+                : EventHandlerSet.Empty;
+        }
+    }
+
+    public readonly record struct WorkspaceEventHandlerAndOptions(Action<EventArgs> Handler, WorkspaceEventOptions Options)
+    {
+    }
+
+    public sealed class EventHandlerSet(ImmutableArray<Registry> registries)
+    {
+        public static readonly EventHandlerSet Empty = new([]);
+        private readonly ImmutableArray<Registry> _registries = registries;
+
+        public static EventHandlerSet Create(WorkspaceEventHandlerAndOptions handlerAndOptions)
+            => new EventHandlerSet([new Registry(handlerAndOptions)]);
+
+        public EventHandlerSet AddHandler(WorkspaceEventHandlerAndOptions handlerAndOptions)
+            => new EventHandlerSet(_registries.Add(new Registry(handlerAndOptions)));
+
+        public EventHandlerSet RemoveHandler(WorkspaceEventHandlerAndOptions handlerAndOptions)
+        {
+            var registry = _registries.FirstOrDefault(static (r, handlerAndOptions) => r.HasHandlerAndOptions(handlerAndOptions), handlerAndOptions);
+
+            if (registry == null)
+                return this;
+
+            // disable this handler (so pending raise events can be squelched)
+            // This does not guarantee no race condition between Raise and Remove but greatly reduces it.
+            registry.Unregister();
+
+            var newRegistries = _registries.Remove(registry);
+
+            return newRegistries.IsEmpty ? Empty : new(newRegistries);
+        }
+
+        public bool HasHandlers
+            => _registries.Any(static r => true);
+
+        public bool HasMatchingOptions(Func<WorkspaceEventOptions, bool> isMatch)
+            => _registries.Any(static (r, hasOptions) => r.HasMatchingOptions(hasOptions), isMatch);
+
+        public void RaiseEvent<TEventArgs>(TEventArgs arg, Func<WorkspaceEventOptions, bool> shouldRaiseEvent)
+            where TEventArgs : EventArgs
+        {
+            foreach (var registry in _registries)
+                registry.RaiseEvent(arg, shouldRaiseEvent);
+        }
+    }
+
+    public sealed class Registry(WorkspaceEventHandlerAndOptions handlerAndOptions)
+    {
+        private readonly WorkspaceEventHandlerAndOptions _handlerAndOptions = handlerAndOptions;
+        private bool _disableHandler = false;
```

Comment:
a potential simpler approach is to make the field mutable, and just assign null to it if unregistered, and use things like `?.Invoke` etc.

Reply:
I don't think I want to go down the path of making a nullable struct in multithreaded code.

Reply:
oh, i didn't realize it was a struct.